In [94]:
input_folder = "./data"

from brat_to_conll import brat_to_conll as b2c

b2c(input_folder, 'dataBIO.txt', 'spacy', 'en_core_web_sm')


Formatting data set from BRAT to CONLL... Warning: brat text and annotation do not match.
	text:  24
	anno: 240
Done.


To solve this warning let check in which file is the mismatch position.

In [129]:
#Read the data
import os

folder_path = "data"
txt_files = [file for file in os.listdir(folder_path) if file.endswith(".txt")]
ann_files = [file for file in os.listdir(folder_path) if file.endswith(".ann")]

files = {'id': [], 'text': [], 'ann': []}
for file in txt_files:
    files['id'].append(file.split('.')[0])
    file_path = os.path.join(folder_path, file)
    with open(file_path, "r") as f:
        files['text'].append(f.read())

for file in ann_files:
    file_path = os.path.join(folder_path, file)
    with open(file_path, "r") as f:
        files['ann'].append(f.read())

In [130]:
#Print the cases where the entity is 240 and the correspondent characters in the text
import codecs

for file in ann_files:
    with codecs.open(f"./data/{file}", 'r', 'UTF-8') as f:
        for line in f.read().splitlines():
            anno = line.split()
            id_anno = anno[0]
            # parse entity
            if id_anno[0] == 'T':
                entity = {}
                entity['id'] = id_anno
                entity['type'] = anno[1]
                entity['start'] = int(anno[2])
                entity['end'] = int(anno[3])
                entity['text'] = ' '.join(anno[4:])

            if entity['text'] == '240':
                print(file)
                print(entity)

                print(files['text'][files['id'].index(file.split('.')[0])][entity['start']:entity['end']])
                
                

31448513.ann
{'id': 'T4', 'type': 'total-participants', 'start': 593, 'end': 596, 'text': '240'}
240
32686337.ann
{'id': 'T5', 'type': 'total-participants', 'start': 357, 'end': 360, 'text': '240'}
240
32686337.ann
{'id': 'T11', 'type': 'total-participants', 'start': 1228, 'end': 1231, 'text': '240'}
240
25935793.ann
{'id': 'T6', 'type': 'control-participants', 'start': 1097, 'end': 1100, 'text': '240'}
240
15023242.ann
{'id': 'T2', 'type': 'total-participants', 'start': 562, 'end': 565, 'text': '240'}
240
15023242.ann
{'id': 'T26', 'type': 'total-participants', 'start': 561, 'end': 564, 'text': '240'}
 24
25068452.ann
{'id': 'T3', 'type': 'total-participants', 'start': 776, 'end': 779, 'text': '240'}
240


The file 15023242.ann is the problem. Start should be 562 and end should be 565. We correct that mistake manually and try again the transformation into BIO format.

In [133]:
b2c(input_folder, 'dataBIO.txt', 'spacy', 'en_core_web_sm')

Formatting data set from BRAT to CONLL... Done.
